diff --git a/src/cmd/acme/text.c b/src/cmd/acme/text.c
index 09422dd..c7d3e3e 100644
--- a/src/cmd/acme/text.c
+++ b/src/cmd/acme/text.c
@@ -760,6 +760,18 @@ texttype(Text *t, Rune r)
 			q0++;
 		textshow(t, q0, q0, TRUE);
 		return;
+	case 0x03:	/* ^C: copy */
+		typecommit(t);
+		cut(t, t, nil, TRUE, FALSE, nil, 0);
+		return;
+	case 0x1A:	/* ^Z: undo */
+	 	typecommit(t);
+		undo(t, nil, nil, TRUE, 0, nil, 0);
+		return;
+	case 0x19: /* ^Y: redo */
+	 	typecommit(t);
+		undo(t, nil, nil, FALSE, 0, nil, 0);
+		return;	
 	case Kcmd+'c':	/* %C: copy */
 		typecommit(t);
 		cut(t, t, nil, TRUE, FALSE, nil, 0);
@@ -796,6 +808,26 @@ texttype(Text *t, Rune r)
 	}
 	/* cut/paste must be done after the seq++/filemark */
 	switch(r){
+	case 0x18:	/* ^X: cut */
+		typecommit(t);
+		if(t->what == Body){
+			seq++;
+			filemark(t->file);
+		}
+		cut(t, t, nil, TRUE, TRUE, nil, 0);
+		textshow(t, t->q0, t->q0, 1);
+		t->iq1 = t->q0;
+		return;
+	case 0x16:	/* ^V: paste */
+		typecommit(t);
+		if(t->what == Body){
+			seq++;
+			filemark(t->file);
+		}
+		paste(t, t, nil, TRUE, FALSE, nil, 0);
+		textshow(t, t->q0, t->q1, 1);
+		t->iq1 = t->q1;
+		return;
 	case Kcmd+'x':	/* %X: cut */
 		typecommit(t);
 		if(t->what == Body){
